google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc_test.cc

```diff
@@ -0,0 +1,462 @@
+
+using ms = std::chrono::milliseconds;
+const unsigned int kFutureWaitMs = 100;
+
+struct FakeRequest {
+  std::string key;
+};
+
+struct FakeResponse {
+  std::string key;
+  std::string value;
+};
+
+using AsyncReaderWriter =
+    ::google::cloud::pubsublite_testing::MockAsyncReaderWriter<FakeRequest,
+                                                               FakeResponse>;
+
+using AsyncReadWriteStreamReturnType =
+    std::unique_ptr<AsyncStreamingReadWriteRpc<FakeRequest, FakeResponse>>;
+
+using ResumableAsyncReadWriteStream = std::unique_ptr<
+    ResumableAsyncStreamingReadWriteRpc<FakeRequest, FakeResponse>>;
+
+const Status kFailStatus = Status(StatusCode::kUnavailable, "Unavailable");
+
+StreamInitializer<FakeRequest, FakeResponse> InitializeInline() {
+  return [](AsyncReadWriteStreamReturnType stream) {
+    return make_ready_future(
+        StatusOr<AsyncReadWriteStreamReturnType>(std::move(stream)));
+  };
+}
+
+std::function<future<void>(std::chrono::milliseconds)> ZeroSleep() {
+  return [](std::chrono::milliseconds) { return make_ready_future(); };
+}
+
+class ResumableAsyncReadWriteStreamingRpcTest : public ::testing::Test {
+ protected:
+  ResumableAsyncReadWriteStreamingRpcTest()
+      : backoff_policy_{std::make_shared<StrictMock<MockBackoffPolicy>>()},
+        stream_{MakeResumableAsyncStreamingReadWriteRpcImpl<FakeRequest,
+                                                            FakeResponse>(
+            retry_policy_factory_.AsStdFunction(), backoff_policy_,
+
+            sleeper_.AsStdFunction(), stream_factory_.AsStdFunction(),
+            initializer_.AsStdFunction())} {}
+
+  StrictMock<MockFunction<future<void>(std::chrono::milliseconds)>> sleeper_;
+  StrictMock<MockFunction<AsyncReadWriteStreamReturnType()>> stream_factory_;
+  StrictMock<MockFunction<std::unique_ptr<MockRetryPolicy>()>>
+      retry_policy_factory_;
+  StrictMock<MockFunction<future<StatusOr<AsyncReadWriteStreamReturnType>>(
+      AsyncReadWriteStreamReturnType)>>
+      initializer_;
+  std::shared_ptr<StrictMock<MockBackoffPolicy>> backoff_policy_;
+  ResumableAsyncReadWriteStream stream_;
+};
+
+TEST_F(ResumableAsyncReadWriteStreamingRpcTest, SingleStartFailureThenGood) {
+  InSequence seq;
+
+  auto stream1 = absl::make_unique<StrictMock<AsyncReaderWriter>>();
+  auto& stream1_ref = *stream1;
+
+  auto mock_retry_policy = absl::make_unique<StrictMock<MockRetryPolicy>>();
+  auto& retry_policy_ref = *mock_retry_policy;
+  EXPECT_CALL(retry_policy_factory_, Call)
+      .WillOnce(Return(ByMove(std::move(mock_retry_policy))));
+
+  auto backoff_policy = absl::make_unique<StrictMock<MockBackoffPolicy>>();
+  auto& backoff_policy_ref = *backoff_policy;
+  EXPECT_CALL(*backoff_policy_, clone)
+      .WillOnce(Return(ByMove(std::move(backoff_policy))));
+
+  EXPECT_CALL(stream_factory_, Call)
+      .WillOnce(Return(ByMove(std::move(stream1))));
+
+  promise<bool> start_promise;
```

Comment:
Just FYI, it does not seem like you need it in these tests. We have a helper class to test asynchronous stuff:

https://github.com/googleapis/google-cloud-cpp/blob/main/google/cloud/testing_util/async_sequencer.h